pkg/events/event_channel.go

```diff
@@ -0,0 +1,74 @@
+/*
+ Licensed to the Apache Software Foundation (ASF) under one
+ or more contributor license agreements.  See the NOTICE file
+ distributed with this work for additional information
+ regarding copyright ownership.  The ASF licenses this file
+ to you under the Apache License, Version 2.0 (the
+ "License"); you may not use this file except in compliance
+ with the License.  You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package events
+
+import (
+	"fmt"
+	"strconv"
+	"sync"
+
+	"github.com/apache/incubator-yunikorn-core/pkg/log"
+)
+
+// TODO should configure the size of the channel
+const eventChannelSize = 100000
+
+// wrapping the channel into a struct - so that the underlying implementation can be changed
+type eventChannel struct {
+	events       chan Event
+	diagCounter  int
+	diagInterval int
+
+	sync.RWMutex
+}
+
+func newEventChannel() *eventChannel {
+	return &eventChannel{
+		events:       make(chan Event, eventChannelSize),
+		diagCounter:  0,
+		diagInterval: eventChannelSize,
+	}
+
+}
+
+func (ec *eventChannel) getNextEvent() (Event, bool) {
+	ec.Lock()
+	defer ec.Unlock()
+
+	select {
+	case msg, ok := <-ec.events:
+		return msg, ok
+	default:
+		return nil, false
+	}
+}
+
+
+func (ec *eventChannel) addEvent(event Event) {
```

Comment:
let's say we have 2 requests, both are pending for some reason.
if the scheduler keeps scheduling for 1 of them, that means the channel will be full of the events just from this one request, and we are losing track of the other one. this looks like a usability issue. 
can we further improve this to store a certain number of events by object? Something like

```
 req1: ev1, ev3, ev4
 req2: ev2, e5
```

Reply:
The `event.GetSource()` is used as key to put the event objects into `EventStore`'s `eventMap`. Currently for all unschedulable requests (for each `AllocationAsk`) we emit an event. In scheduler's while loop if can't suffice a request, we emit an event and get to the next one, where we also emit an event. Because the asks are different, every events will be stored.

I see one problem though that is when we have multiple events for the same request. I don't really see this as a problem now because in the scheduler when we bump into an issue (like resource shortage) the whole request is skipped - no additional check is done. If you can provide a reasonable example when this is the case I can refactor the backend for the `EventStore`.